src/neuroglancer/datasource/deepzoom/frontend.ts

```diff
@@ -0,0 +1,304 @@
+  resolution: Float64Array;
+  voxelOffset: Float32Array;
+  size: Float32Array;
+  chunkSizes: Uint32Array[];
+  constructor(obj: any, numChannels: number) {
+    verifyObject(obj);
+    const rank = (numChannels === 1) ? 3 : 4;
+    const resolution = this.resolution = new Float64Array(rank);
+    const voxelOffset = this.voxelOffset = new Float32Array(rank);
+    const size = this.size = new Float32Array(rank);
+    if (rank === 4) {
+      resolution[3] = 1;
+      size[3] = numChannels;
+    }
+    verifyObjectProperty(
+        obj, 'resolution',
+        x => parseFixedLengthArray(resolution.subarray(0, 3), x, verifyFinitePositiveFloat));
+    verifyOptionalObjectProperty(
+        obj, 'voxel_offset', x => parseFixedLengthArray(voxelOffset.subarray(0, 3), x, verifyInt));
+    verifyObjectProperty(
+        obj, 'size', x => parseFixedLengthArray(size.subarray(0, 3), x, verifyPositiveInt));
+    this.chunkSizes = verifyObjectProperty(
+        obj, 'chunk_sizes', x => parseArray(x, y => {
+                              const chunkSize = new Uint32Array(rank);
+                              if (rank === 4) chunkSize[3] = numChannels;
+                              parseFixedLengthArray(chunkSize.subarray(0, 3), y, verifyPositiveInt);
+                              return chunkSize;
+                            }));
+    if (this.chunkSizes.length === 0) {
+      throw new Error('No chunk sizes specified.');
+    }
+    this.encoding =
+        verifyObjectProperty(obj, 'encoding', x => verifyEnumString(x, ImageTileEncoding));
+    this.key = verifyObjectProperty(obj, 'key', verifyString);
+  }
+}
+
+/*export*/ interface PyramidalImageInfo {
+  dataType: DataType;
+  volumeType: VolumeType;
+  scales: ScaleInfo[];
+  modelSpace: CoordinateSpace;
+  overlap: number;
+  tilesize: number;
+}
+
+/*export*/ function buildPyramidalImageInfo(metadata: DZIMetaData): PyramidalImageInfo {
+  const {width, height, tilesize, overlap, format} = metadata;
+  const dataType = DataType.UINT8;
+  const numChannels = 3;
+  const volumeType = VolumeType.IMAGE;
+  const scaleInfos = new Array<ScaleInfo>();
+  let w = width, h = height;
+  let maxlevel = Math.ceil(Math.log2(Math.max(w, h)));
+  do {
+    const lvl = scaleInfos.length;
+    const res = 1 << lvl;
+    scaleInfos.push(new ScaleInfo(
+        {
+          key: (maxlevel - lvl).toString(),
+          size: [w, h, 1],
+          resolution: [res, res, res],
+          chunk_sizes: [[tilesize, tilesize, 1]],
+          encoding: format
+        },
+        numChannels));
+    w = Math.ceil(w / 2);
+    h = Math.ceil(h / 2);
+  } while (w > 1 || h > 1);
+
+  if (scaleInfos.length === 0) throw new Error('Expected at least one scale');
+  const baseScale = scaleInfos[0];
+  const rank = 4;
+  const scales = new Float64Array(rank);
+  const lowerBounds = new Float64Array(rank);
+  const upperBounds = new Float64Array(rank);
+  const names = ['x', 'y', 'z'];
+  const units = ['m', 'm', 'm'];
```

Comment:
The unit can be the empty string `''` if the unit is unknown.

Reply:
I think I've done that at a point and it triggered something about the 1e9 factor, so target dimensions appeared as `<width>`, `<height>`, `1000000000` (or `1e9`). For this one I will check the new hidden dimension feature anyway.
Side note: Deep Zoom images have no physical dimensions specified inside them. I just left the `m`s in place. Tried with `px`, but the target units still remained `m`, and upon scaling it introduced some fancy units, like nanopixels (`npx`).

Reply:
I think the issue you observed was simply due to using an old neuroglancer json state with the updated code, e.g. due to using `dev-server`.  The global coordinate space units and scale are part of the json state.

Reply:
Note: The "hidden" scale property added recently for precomputed is something entirely different, not related to this.  That simply allows some of the scale levels that are specified in the precomputed metadata to be ignored.

Reply:
Yes, this works after all, but has extreme side effects on the other views (makes the plane infinitely thick).

As of https://github.com/google/neuroglancer/commit/ceca06e04a75b5eb39eb435dbd3058decefb5d85 the block starts from line 64, and

    const rank = 4;
    const scales = Float64Array.of(1 / 1e9, 1 / 1e9, 1, 1);
    const lowerBounds = new Float64Array(rank);
    const upperBounds = Float64Array.of(width, height, 1, 3);
    const names = ['x', 'y', 'z', 'c^'];
    const units = ['m', 'm', '', ''];

would be the "strange" variant of it I'm talking about.

Reply:
Yes, by default a 2-d volume will be "infinitely" thick in the third dimension.  But why do you care about other views?  Surely for a 2d dataset you would just want the xy view?  You can through the layer's coordinate transform project the 2d dataset to a 2d linear subspace of a higher-dimensional space.  That is done in the UI via the "+V" button at the bottom of the coordinate transform matrix.

Reply:
Perhaps this could be an issue if overlaying a 2d image onto a 3-D volume.

Reply:
Yes, it depends on what you want to do --- the default behavior is to "broadcast" the volume along any other dimensions, i.e. as if it is one infinitely thick voxel in that dimension, which is often the desired result when mixing 2d/3d/4d datasets.  But as noted, you can also project to a 2d subspace of a 3d coordinate space, so that it will display as a single "slice" in the 3d volume.

Reply:
Can you say a bit more what you find disturbing?  The precomputed volume format is limited to 3d multichannel, just as deepzoom is limited to 2d multichannel.  But e.g. the zarr and n5 data sources allow an arbitrary number of dimensions.

Reply:
> Can you say a bit more what you find disturbing?
1. I made changes which were systematic, but I don't actually know what they interface with
2. *and* the `xz` and `yz` views started to behave differently. I understand they don't matter, but I don't understand why they became different. I'd expect `yz` to produce the same striped pattern, just perhaps horizontally (though in fact that would be `zy`).

And it does, after adding a fake "axis": [strange](https://www.nesys.uio.no/NG+DZI/#!%7B%22dimensions%22:%7B%22x%22:%5B1e-9%2C%22m%22%5D%2C%22y%22:%5B1e-9%2C%22m%22%5D%7D%2C%22position%22:%5B26327.98828125%2C18448.6015625%5D%2C%22crossSectionScale%22:141.17496392147683%2C%22projectionScale%22:65536%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%22deepzoom://https://data-proxy.ebrains.eu/api/v1/buckets/localizoom/14122_mPPC_BDA_s186.tif/14122_mPPC_BDA_s186.dzi%22%2C%22tab%22:%22source%22%2C%22shader%22:%22void%20main%20%28%29%20%7B%5Cn%20%20emitRGB%28vec3%28toNormalized%28getDataValue%280%29%29%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20toNormalized%28getDataValue%281%29%29%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20toNormalized%28getDataValue%282%29%29%29%29%3B%5Cn%7D%22%2C%22channelDimensions%22:%7B%22c%5E%22:%5B1%2C%22%22%5D%7D%2C%22name%22:%2214122_mPPC_BDA_s186.dzi%22%7D%5D%2C%22selectedLayer%22:%7B%22visible%22:true%2C%22layer%22:%2214122_mPPC_BDA_s186.dzi%22%7D%2C%22layout%22:%224panel%22%7D) vs. [expected](https://www.nesys.uio.no/NG+DZI/#!%7B%22dimensions%22:%7B%22x%22:%5B1e-9%2C%22m%22%5D%2C%22y%22:%5B1e-9%2C%22m%22%5D%2C%22axis%22:%5B1%2C%22%22%5D%7D%2C%22position%22:%5B26327.98828125%2C18448.6015625%2C0.5%5D%2C%22crossSectionScale%22:141.17496392147683%2C%22projectionScale%22:65536%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%7B%22url%22:%22deepzoom://https://data-proxy.ebrains.eu/api/v1/buckets/localizoom/14122_mPPC_BDA_s186.tif/14122_mPPC_BDA_s186.dzi%22%2C%22transform%22:%7B%22sourceRank%22:3%2C%22outputDimensions%22:%7B%22x%22:%5B1e-9%2C%22m%22%5D%2C%22y%22:%5B1e-9%2C%22m%22%5D%2C%22c%5E%22:%5B1%2C%22%22%5D%2C%22axis%22:%5B1%2C%22%22%5D%7D%2C%22inputDimensions%22:%7B%22x%22:%5B1e-9%2C%22m%22%5D%2C%22y%22:%5B1e-9%2C%22m%22%5D%2C%22c%5E%22:%5B1%2C%22%22%5D%2C%22%22:%5B1%2C%22%22%5D%7D%7D%7D%2C%22tab%22:%22source%22%2C%22shader%22:%22void%20main%20%28%29%20%7B%5Cn%20%20emitRGB%28vec3%28toNormalized%28getDataValue%280%29%29%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20toNormalized%28getDataValue%281%29%29%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20toNormalized%28getDataValue%282%29%29%29%29%3B%5Cn%7D%22%2C%22channelDimensions%22:%7B%22c%5E%22:%5B1%2C%22%22%5D%7D%2C%22name%22:%2214122_mPPC_BDA_s186.dzi%22%7D%5D%2C%22selectedLayer%22:%7B%22visible%22:true%2C%22layer%22:%2214122_mPPC_BDA_s186.dzi%22%7D%2C%22layout%22:%224panel%22%7D)